### PR TITLE
Added option to pass password for pem file

### PR DESCRIPTION
**Pull request by waqar-baig**
Added option to pass password for pem file

### CODE REVIEW

**Comment by vanso-waqar** [comment]
Is it ready to merge?

**Comment by calonso** [comment]
Yeah, this looks good, just update the documentation of the affected methods and add a small spec that checks that the password given at the pusher is then received by `OpenSSL::PKey::RSA.new cert` call in `APNSConnection`

**Comment by calonso** [comment]
How are you doing with this? Do you need any help?

**Comment by waqar-baig** [comment]
Need some time. Will do it. 


     On Monday, 27 April 2015 7:31 PM, Carlos Alonso <notifications@github.com> wrote:
   

 How are you doing with this? Do you need any help?—

**Comment by calonso** [comment]
Ok, cool, just let me know if you need a hand.

**Comment by waqar-baig** [comment]
I believe it is ready to merge :)

**Comment by calonso** [comment]
Please, rebase this branch to have it up to date with master.

File: lib/ruby-push-notifications/apns/apns_notification.rb [hunk 1]
```diff
@@ -67,7 +67,7 @@ def device_token(token)
       # @return [String]. Binary representation of the notification's payload.
       def payload
         @encoded_payload ||= -> {
-            json = JSON.dump(@data).force_encoding 'ascii-8bit'
+            json = (@data.respond_to?(:to_json) ? @data.to_json : JSON.dump(@data)).force_encoding 'ascii-8bit'
```
Comment (calonso) [inline lib/ruby-push-notifications/apns/apns_notification.rb]: What do we need this for?

**Comment by calonso** [comment]
@waqar-baig left a couple of comments, good work apart for this. Thanks!

File: lib/ruby-push-notifications/apns/apns_notification.rb (hunk 1 above)
Comment (waqar-baig) [inline lib/ruby-push-notifications/apns/apns_notification.rb]: When working with rails activerecord object `JSON.dump` has issues. So better to use `.to_json`

Comment (calonso) [inline lib/ruby-push-notifications/apns/apns_notification.rb]: I'm curious, which kind of issues?

Comment (calonso) [inline lib/ruby-push-notifications/apns/apns_notification.rb]: @waqar-baig. I just want to know the issues you have faced, to be sure we really need this here.

Apart from this, are you working on the documentation?

**Comment by calonso** [comment]
@waqar-baig How are you doing with this? Do you need any help?

**Comment by xykox** [comment]
Hi all,
will the pull request be merged  ?

**Comment by calonso** [comment]
Yeah, give me a couple of hours

**Comment by calonso** [comment]
Actually it needs to be @waqar-baig who finishes it as the code is in his fork.
@waqar-baig Are you up for finishing this?
TODO list:
* The JSON dump of the APNS notification has to be in ascii-8bit and I don't think `to_json`encodes using it
* Docs update to describe the new argument.

Apart from this small details this is a good feature. Cheers!

**Comment by xykox** [comment]
thx :)

**Comment by waqar-baig** [comment]
I had issues serializing deep ActiveRecord objects with JSON.dump.
So, I had changed it to #to_json.
Also using this same fork in my production without issues.
You guys can give it a try and merge if it works for you :)

**Comment by calonso** [comment]
@waqar-baig You're right, I didn't read the force_encoding properly.

Could you please update the README file to include the new parameter documentation? Then I'll merge it straight away.

BTW, very glad to hear you're using the gem in production nicely! :)

**Comment by waqar-baig** [comment]
I have already updated the apns.rb example for the same.
I hope you are looking for the same.
Cause readme file doesn't have code snippets.

File: spec/ruby-push-notifications/apns/apns_notification_spec.rb [hunk 2]
```diff
@@ -16,6 +16,7 @@ module APNS
       end
 
       it 'caches the payload' do
+        allow(data).to receive(:respond_to?).and_return(false)
```
Comment (calonso) [inline spec/ruby-push-notifications/apns/apns_notification_spec.rb]: Allowing to receive `respond_to` is probably too generic. Could you please add `with(:to_json)` to limit the stubbing?

**Comment by calonso** [comment]
Sorry @waqar-baig I thought there were some snippets in README file. Sorry about that. 
Just left a couple of minor comments about tests and it should be ready to merge.
Great work!

**Comment by calonso** [comment]
That's really cool. Thanks @waqar-baig for your effort.

**Comment by calonso** [comment]
Travis build failed because the `Gemfile.lock`has the previous version referenced. See: [https://github.com/calonso/ruby-push-notifications/blob/master/Gemfile.lock#L4](https://github.com/calonso/ruby-push-notifications/blob/master/Gemfile.lock#L4)
Please, run `bundle update` to update the reference and that should fix it.

**Comment by calonso** [comment]
@waqar-baig Perfect, thanks!!

**Pull request merged**